File: worker/src/protocol.ts
/**
 * Trusted-DNS Protocol Constants and Types
 *
 * Defines the binary protocol structures, message types, error codes,
 * and key derivation utilities shared across Worker modules.
 */

// ─── Protocol Version ───────────────────────────────────────────────
export const PROTOCOL_VERSION = 0x01;

// ─── Message Types ──────────────────────────────────────────────────
export const MSG_BOOTSTRAP_REQ  = 0x01;
export const MSG_BOOTSTRAP_RESP = 0x02;
export const MSG_QUERY_REQ      = 0x03;
export const MSG_QUERY_RESP     = 0x04;
export const MSG_REFRESH_REQ    = 0x05;
export const MSG_REFRESH_RESP   = 0x06;
export const MSG_ERROR_RESP     = 0x7f;

// ─── Error Codes ────────────────────────────────────────────────────
export const ERR_BAD_VERSION      = 0x01;
export const ERR_BAD_TYPE         = 0x02;
export const ERR_BAD_TICKET       = 0x03;
export const ERR_EXPIRED          = 0x04;
export const ERR_OLD_GENERATION   = 0x05;
export const ERR_REPLAY_SUSPECTED = 0x06;
export const ERR_DECRYPT_FAILED   = 0x07;
export const ERR_UPSTREAM_FAILURE = 0x08;
export const ERR_RATE_LIMITED     = 0x09;
export const ERR_INTERNAL         = 0x0a;

// ─── Fixed Sizes ────────────────────────────────────────────────────
export const HEADER_SIZE        = 32;
export const CLIENT_ID_SIZE     = 32;
export const CLIENT_ID_PREFIX   = 8;
export const NONCE_SIZE         = 12;
export const TAG_SIZE           = 16;
export const TICKET_TAG_SIZE    = 16;
export const RESUME_SEED_SIZE   = 32;
export const REFRESH_SEED_SIZE  = 32;
export const REFRESH_NONCE_SIZE = 16;
export const BOOT_NONCE_SIZE    = 16;

// ─── Default Policy Values ──────────────────────────────────────────
export const DEFAULT_TICKETS_PER_BUNDLE  = 5;
export const DEFAULT_QUERIES_PER_TICKET  = 200;
export const DEFAULT_QUERIES_PER_BUNDLE  = 1000;
export const DEFAULT_MAX_CLOCK_SKEW_MS   = 300_000;
export const DEFAULT_ANTI_REPLAY_WINDOW  = 64;
export const DEFAULT_TICKET_LIFETIME_MS  = 3_600_000; // 1 hour

// ─── Interfaces ─────────────────────────────────────────────────────

/** 32-byte fixed protocol header (big-endian) */
export interface ProtocolHeader {
  ver: number;             // 1 byte
  msgType: number;         // 1 byte
  flags: number;           // 2 bytes
  clientIdPrefix: Uint8Array; // 8 bytes
  bundleGen: bigint;       // 8 bytes (u64)
  ticketId: number;        // 2 bytes
  seq: number;             // 4 bytes
  payloadLen: number;      // 4 bytes
  headerMac: number;       // 2 bytes
}

export interface Policy {
  ticketsPerBundle: number;
  queriesPerTicket: number;
  queriesPerBundle: number;
  maxClockSkewMs: number;
  antiReplayWindow: number;
  ticketLifetimeMs: number;
}

export interface SessionTicket {
  ticketId: number;        // u16
  slot: number;            // u8
  reserved: number;        // u8
  clientId: Uint8Array;    // 32 bytes
  bundleGen: bigint;       // u64
  notBeforeMs: bigint;     // u64
  notAfterMs: bigint;      // u64
  queryBudget: number;     // u16
  counterBase: number;     // u32
  resumeSeed: Uint8Array;  // 32 bytes
  ticketTag: Uint8Array;   // 16 bytes
}

export interface RefreshTicket {
  clientId: Uint8Array;    // 32 bytes
  bundleGen: bigint;       // u64
  notBeforeMs: bigint;     // u64
  notAfterMs: bigint;      // u64
  rotateAfterQueries: number; // u16
  refreshNonce: Uint8Array;   // 16 bytes
  refreshSeed: Uint8Array;    // 32 bytes
  refreshTag: Uint8Array;     // 16 bytes
}

export interface KeyBundle {
  bundleGen: bigint;
  issuedAtMs: bigint;
  expireAtMs: bigint;
  workerKid: number;
  policy: Policy;
  sessionTickets: SessionTicket[];
  refreshTicket: RefreshTicket;
}

// ─── Header Encoding / Decoding ─────────────────────────────────────

export function encodeHeader(h: ProtocolHeader): Uint8Array {
  const buf = new ArrayBuffer(HEADER_SIZE);
  const dv = new DataView(buf);
  const u8 = new Uint8Array(buf);

  dv.setUint8(0, h.ver);
  dv.setUint8(1, h.msgType);
  dv.setUint16(2, h.flags, false);
  u8.set(h.clientIdPrefix.subarray(0, 8), 4);
  dv.setBigUint64(12, h.bundleGen, false);
  dv.setUint16(20, h.ticketId, false);
  dv.setUint32(22, h.seq, false);
  dv.setUint32(26, h.payloadLen, false);
  dv.setUint16(30, h.headerMac, false);

  return u8;
}

export function decodeHeader(data: Uint8Array): ProtocolHeader {
  if (data.length < HEADER_SIZE) {
    throw new Error(`Header too short: ${data.length} < ${HEADER_SIZE}`);
  }
  const dv = new DataView(data.buffer, data.byteOffset, data.byteLength);

  return {
    ver: dv.getUint8(0),
    msgType: dv.getUint8(1),
    flags: dv.getUint16(2, false),
    clientIdPrefix: data.slice(4, 12),
    bundleGen: dv.getBigUint64(12, false),
    ticketId: dv.getUint16(20, false),
    seq: dv.getUint32(22, false),
    payloadLen: dv.getUint32(26, false),
    headerMac: dv.getUint16(30, false),
  };
}

// ─── Session Ticket Serialization ───────────────────────────────────

export const SESSION_TICKET_SIZE = 2 + 1 + 1 + 32 + 8 + 8 + 8 + 2 + 4 + 32 + 16; // = 114

export function encodeSessionTicket(t: SessionTicket): Uint8Array {
  const buf = new ArrayBuffer(SESSION_TICKET_SIZE);
  const dv = new DataView(buf);
  const u8 = new Uint8Array(buf);
  let off = 0;

  dv.setUint16(off, t.ticketId, false); off += 2;
  dv.setUint8(off, t.slot); off += 1;
  dv.setUint8(off, t.reserved); off += 1;
  u8.set(t.clientId.subarray(0, 32), off); off += 32;
  dv.setBigUint64(off, t.bundleGen, false); off += 8;
  dv.setBigUint64(off, t.notBeforeMs, false); off += 8;
  dv.setBigUint64(off, t.notAfterMs, false); off += 8;
  dv.setUint16(off, t.queryBudget, false); off += 2;
  dv.setUint32(off, t.counterBase, false); off += 4;
  u8.set(t.resumeSeed.subarray(0, 32), off); off += 32;
  u8.set(t.ticketTag.subarray(0, 16), off);

  return u8;
}

export function decodeSessionTicket(data: Uint8Array): SessionTicket {
  if (data.length < SESSION_TICKET_SIZE) {
    throw new Error(`SessionTicket too short: ${data.length}`);
  }
  const dv = new DataView(data.buffer, data.byteOffset, data.byteLength);
  let off = 0;

  const ticketId = dv.getUint16(off, false); off += 2;
  const slot = dv.getUint8(off); off += 1;
  const reserved = dv.getUint8(off); off += 1;
  const clientId = data.slice(off, off + 32); off += 32;
  const bundleGen = dv.getBigUint64(off, false); off += 8;
  const notBeforeMs = dv.getBigUint64(off, false); off += 8;
  const notAfterMs = dv.getBigUint64(off, false); off += 8;
  const queryBudget = dv.getUint16(off, false); off += 2;
  const counterBase = dv.getUint32(off, false); off += 4;
  const resumeSeed = data.slice(off, off + 32); off += 32;
  const ticketTag = data.slice(off, off + 16);

  return {
    ticketId, slot, reserved, clientId, bundleGen,
    notBeforeMs, notAfterMs, queryBudget, counterBase,
    resumeSeed, ticketTag,
  };
}

// ─── Refresh Ticket Serialization ───────────────────────────────────

export const REFRESH_TICKET_SIZE = 32 + 8 + 8 + 8 + 2 + 16 + 32 + 16; // = 122

export function encodeRefreshTicket(t: RefreshTicket): Uint8Array {
  const buf = new ArrayBuffer(REFRESH_TICKET_SIZE);
  const dv = new DataView(buf);
  const u8 = new Uint8Array(buf);
  let off = 0;

  u8.set(t.clientId.subarray(0, 32), off); off += 32;
  dv.setBigUint64(off, t.bundleGen, false); off += 8;
  dv.setBigUint64(off, t.notBeforeMs, false); off += 8;
  dv.setBigUint64(off, t.notAfterMs, false); off += 8;
  dv.setUint16(off, t.rotateAfterQueries, false); off += 2;
  u8.set(t.refreshNonce.subarray(0, 16), off); off += 16;
  u8.set(t.refreshSeed.subarray(0, 32), off); off += 32;
  u8.set(t.refreshTag.subarray(0, 16), off);

  return u8;
}

export function decodeRefreshTicket(data: Uint8Array): RefreshTicket {
  if (data.length < REFRESH_TICKET_SIZE) {
    throw new Error(`RefreshTicket too short: ${data.length}`);
  }
  const dv = new DataView(data.buffer, data.byteOffset, data.byteLength);
  let off = 0;

  const clientId = data.slice(off, off + 32); off += 32;
  const bundleGen = dv.getBigUint64(off, false); off += 8;
  const notBeforeMs = dv.getBigUint64(off, false); off += 8;
  const notAfterMs = dv.getBigUint64(off, false); off += 8;
  const rotateAfterQueries = dv.getUint16(off, false); off += 2;
  const refreshNonce = data.slice(off, off + 16); off += 16;
  const refreshSeed = data.slice(off, off + 32); off += 32;
  const refreshTag = data.slice(off, off + 16);

  return {
    clientId, bundleGen, notBeforeMs, notAfterMs,
    rotateAfterQueries, refreshNonce, refreshSeed, refreshTag,
  };
}

// ─── Error Response Builder ─────────────────────────────────────────

export function buildErrorResponse(
  clientIdPrefix: Uint8Array,
  bundleGen: bigint,
  errorCode: number,
  detail?: string,
): Uint8Array {
  const detailBytes = detail
    ? new TextEncoder().encode(detail.substring(0, 128))
    : new Uint8Array(0);
  const payloadLen = 1 + detailBytes.length;

  const header = encodeHeader({
    ver: PROTOCOL_VERSION,
    msgType: MSG_ERROR_RESP,
    flags: 0,
    clientIdPrefix,
    bundleGen,
    ticketId: 0,
    seq: 0,
    payloadLen,
    headerMac: 0,
  });

  const result = new Uint8Array(HEADER_SIZE + payloadLen);
  result.set(header, 0);
  result[HEADER_SIZE] = errorCode;
  if (detailBytes.length > 0) {
    result.set(detailBytes, HEADER_SIZE + 1);
  }
  return result;
}

File: docker/internal/protocol/protocol.go
// Package protocol defines the Trusted-DNS binary protocol structures,
// message types, error codes, and serialization/deserialization routines.
package protocol

import (
	"encoding/binary"
	"errors"
	"fmt"
)

// Protocol version
const ProtocolVersion = 0x01

// Message types
const (
	MsgBootstrapReq  = 0x01
	MsgBootstrapResp = 0x02
	MsgQueryReq      = 0x03
	MsgQueryResp     = 0x04
	MsgRefreshReq    = 0x05
	MsgRefreshResp   = 0x06
	MsgErrorResp     = 0x7F
)

// Error codes
const (
	ErrBadVersion     = 0x01
	ErrBadType        = 0x02
	ErrBadTicket      = 0x03
	ErrExpired        = 0x04
	ErrOldGeneration  = 0x05
	ErrReplaySuspect  = 0x06
	ErrDecryptFailed  = 0x07
	ErrUpstreamFail   = 0x08
	ErrRateLimited    = 0x09
	ErrInternal       = 0x0A
)

// Fixed sizes
const (
	HeaderSize       = 32
	ClientIDSize     = 32
	ClientIDPrefix   = 8
	NonceSize        = 12
	TagSize          = 16
	TicketTagSize    = 16
	ResumeSeedSize   = 32
	RefreshSeedSize  = 32
	RefreshNonceSize = 16
	BootNonceSize    = 16
)

// Default policy values
const (
	DefaultTicketsPerBundle = 5
	DefaultQueriesPerTicket = 200
	DefaultQueriesPerBundle = 1000
	DefaultMaxClockSkewMs   = 300000
	DefaultAntiReplayWindow = 64
	DefaultTicketLifetimeMs = 3600000
)

// Header represents the 32-byte fixed protocol header.
type Header struct {
	Ver            uint8
	MsgType        uint8
	Flags          uint16
	ClientIDPrefix [8]byte
	BundleGen      uint64
	TicketID       uint16
	Seq            uint32
	PayloadLen     uint32
	HeaderMAC      uint16
}

// EncodeHeader serializes a Header to 32 bytes (big-endian).
func EncodeHeader(h *Header) []byte {
	buf := make([]byte, HeaderSize)
	buf[0] = h.Ver
	buf[1] = h.MsgType
	binary.BigEndian.PutUint16(buf[2:4], h.Flags)
	copy(buf[4:12], h.ClientIDPrefix[:])
	binary.BigEndian.PutUint64(buf[12:20], h.BundleGen)
	binary.BigEndian.PutUint16(buf[20:22], h.TicketID)
	binary.BigEndian.PutUint32(buf[22:26], h.Seq)
	binary.BigEndian.PutUint32(buf[26:30], h.PayloadLen)
	binary.BigEndian.PutUint16(buf[30:32], h.HeaderMAC)
	return buf
}

// DecodeHeader deserializes 32 bytes into a Header.
func DecodeHeader(data []byte) (*Header, error) {
	if len(data) < HeaderSize {
		return nil, fmt.Errorf("header too short: %d < %d", len(data), HeaderSize)
	}
	h := &Header{
		Ver:     data[0],
		MsgType: data[1],
		Flags:   binary.BigEndian.Uint16(data[2:4]),
	}
	copy(h.ClientIDPrefix[:], data[4:12])
	h.BundleGen = binary.BigEndian.Uint64(data[12:20])
	h.TicketID = binary.BigEndian.Uint16(data[20:22])
	h.Seq = binary.BigEndian.Uint32(data[22:26])
	h.PayloadLen = binary.BigEndian.Uint32(data[26:30])
	h.HeaderMAC = binary.BigEndian.Uint16(data[30:32])
	return h, nil
}

// Policy holds the session policy parameters.
type Policy struct {
	TicketsPerBundle uint8
	QueriesPerTicket uint16
	QueriesPerBundle uint16
	MaxClockSkewMs   uint32
	AntiReplayWindow uint16
	TicketLifetimeMs uint64
}

// DefaultPolicy returns the default policy.
func DefaultPolicy() Policy {
	return Policy{
		TicketsPerBundle: DefaultTicketsPerBundle,
		QueriesPerTicket: DefaultQueriesPerTicket,
		QueriesPerBundle: DefaultQueriesPerBundle,
		MaxClockSkewMs:   DefaultMaxClockSkewMs,
		AntiReplayWindow: DefaultAntiReplayWindow,
		TicketLifetimeMs: DefaultTicketLifetimeMs,
	}
}

// SessionTicket represents a single session ticket.
const SessionTicketSize = 2 + 1 + 1 + 32 + 8 + 8 + 8 + 2 + 4 + 32 + 16 // = 114

type SessionTicket struct {
	TicketID    uint16
	Slot        uint8
	Reserved    uint8
	ClientID    [32]byte
	BundleGen   uint64
	NotBeforeMs uint64
	NotAfterMs  uint64
	QueryBudget uint16
	CounterBase uint32
	ResumeSeed  [32]byte
	TicketTag   [16]byte
}

// EncodeSessionTicket serializes a SessionTicket.
func EncodeSessionTicket(t *SessionTicket) []byte {
	buf := make([]byte, SessionTicketSize)
	off := 0
	binary.BigEndian.PutUint16(buf[off:], t.TicketID); off += 2
	buf[off] = t.Slot; off++
	buf[off] = t.Reserved; off++
	copy(buf[off:], t.ClientID[:]); off += 32
	binary.BigEndian.PutUint64(buf[off:], t.BundleGen); off += 8
	binary.BigEndian.PutUint64(buf[off:], t.NotBeforeMs); off += 8
	binary.BigEndian.PutUint64(buf[off:], t.NotAfterMs); off += 8
	binary.BigEndian.PutUint16(buf[off:], t.QueryBudget); off += 2
	binary.BigEndian.PutUint32(buf[off:], t.CounterBase); off += 4
	copy(buf[off:], t.ResumeSeed[:]); off += 32
	copy(buf[off:], t.TicketTag[:])
	return buf
}

// DecodeSessionTicket deserializes a SessionTicket.
func DecodeSessionTicket(data []byte) (*SessionTicket, error) {
	if len(data) < SessionTicketSize {
		return nil, fmt.Errorf("session ticket too short: %d", len(data))
	}
	t := &SessionTicket{}
	off := 0
	t.TicketID = binary.BigEndian.Uint16(data[off:]); off += 2
	t.Slot = data[off]; off++
	t.Reserved = data[off]; off++
	copy(t.ClientID[:], data[off:off+32]); off += 32
	t.BundleGen = binary.BigEndian.Uint64(data[off:]); off += 8
	t.NotBeforeMs = binary.BigEndian.Uint64(data[off:]); off += 8
	t.NotAfterMs = binary.BigEndian.Uint64(data[off:]); off += 8
	t.QueryBudget = binary.BigEndian.Uint16(data[off:]); off += 2
	t.CounterBase = binary.BigEndian.Uint32(data[off:]); off += 4
	copy(t.ResumeSeed[:], data[off:off+32]); off += 32
	copy(t.TicketTag[:], data[off:off+16])
	return t, nil
}

// RefreshTicket represents a refresh ticket.
const RefreshTicketSize = 32 + 8 + 8 + 8 + 2 + 16 + 32 + 16 // = 122

type RefreshTicket struct {
	ClientID           [32]byte
	BundleGen          uint64
	NotBeforeMs        uint64
	NotAfterMs         uint64
	RotateAfterQueries uint16
	RefreshNonce       [16]byte
	RefreshSeed        [32]byte
	RefreshTag         [16]byte
}

// EncodeRefreshTicket serializes a RefreshTicket.
func EncodeRefreshTicket(t *RefreshTicket) []byte {
	buf := make([]byte, RefreshTicketSize)
	off := 0
	copy(buf[off:], t.ClientID[:]); off += 32
	binary.BigEndian.PutUint64(buf[off:], t.BundleGen); off += 8
	binary.BigEndian.PutUint64(buf[off:], t.NotBeforeMs); off += 8
	binary.BigEndian.PutUint64(buf[off:], t.NotAfterMs); off += 8
	binary.BigEndian.PutUint16(buf[off:], t.RotateAfterQueries); off += 2
	copy(buf[off:], t.RefreshNonce[:]); off += 16
	copy(buf[off:], t.RefreshSeed[:]); off += 32
	copy(buf[off:], t.RefreshTag[:])
	return buf
}

// DecodeRefreshTicket deserializes a RefreshTicket.
func DecodeRefreshTicket(data []byte) (*RefreshTicket, error) {
	if len(data) < RefreshTicketSize {
		return nil, fmt.Errorf("refresh ticket too short: %d", len(data))
	}
	t := &RefreshTicket{}
	off := 0
	copy(t.ClientID[:], data[off:off+32]); off += 32
	t.BundleGen = binary.BigEndian.Uint64(data[off:]); off += 8
	t.NotBeforeMs = binary.BigEndian.Uint64(data[off:]); off += 8
	t.NotAfterMs = binary.BigEndian.Uint64(data[off:]); off += 8
	t.RotateAfterQueries = binary.BigEndian.Uint16(data[off:]); off += 2
	copy(t.RefreshNonce[:], data[off:off+16]); off += 16
	copy(t.RefreshSeed[:], data[off:off+32]); off += 32
	copy(t.RefreshTag[:], data[off:off+16])
	return t, nil
}

// KeyBundle holds the complete session material set.
type KeyBundle struct {
	BundleGen      uint64
	IssuedAtMs     uint64
	ExpireAtMs     uint64
	WorkerKID      uint16
	Policy         Policy
	SessionTickets []*SessionTicket
	RefreshTicket  *RefreshTicket
}

// ErrorResponse represents a protocol error.
type ErrorResponse struct {
	Code   uint8
	Detail string
}

func (e *ErrorResponse) Error() string {
	return fmt.Sprintf("protocol error 0x%02x: %s", e.Code, e.Detail)
}

// ParseErrorResponse extracts error info from an ERROR_RESP payload.
func ParseErrorResponse(payload []byte) *ErrorResponse {
	if len(payload) < 1 {
		return &ErrorResponse{Code: ErrInternal, Detail: "empty error payload"}
	}
	detail := ""
	if len(payload) > 1 {
		detail = string(payload[1:])
	}
	return &ErrorResponse{Code: payload[0], Detail: detail}
}

// SerializeKeyBundle serializes a KeyBundle for decryption on the Docker side.
func SerializeKeyBundle(b *KeyBundle) []byte {
	policySize := 1 + 1 + 2 + 2 + 4 + 2 + 8 // = 20
	totalSize := 8 + 8 + 8 + 2 + policySize +
		len(b.SessionTickets)*SessionTicketSize + RefreshTicketSize

	buf := make([]byte, totalSize)
	off := 0

	binary.BigEndian.PutUint64(buf[off:], b.BundleGen); off += 8
	binary.BigEndian.PutUint64(buf[off:], b.IssuedAtMs); off += 8
	binary.BigEndian.PutUint64(buf[off:], b.ExpireAtMs); off += 8
	binary.BigEndian.PutUint16(buf[off:], b.WorkerKID); off += 2

	// Policy
	buf[off] = b.Policy.TicketsPerBundle; off++
	buf[off] = 0; off++ // padding
	binary.BigEndian.PutUint16(buf[off:], b.Policy.QueriesPerTicket); off += 2
	binary.BigEndian.PutUint16(buf[off:], b.Policy.QueriesPerBundle); off += 2
	binary.BigEndian.PutUint32(buf[off:], b.Policy.MaxClockSkewMs); off += 4
	binary.BigEndian.PutUint16(buf[off:], b.Policy.AntiReplayWindow); off += 2
	binary.BigEndian.PutUint64(buf[off:], b.Policy.TicketLifetimeMs); off += 8

	for _, t := range b.SessionTickets {
		copy(buf[off:], EncodeSessionTicket(t))
		off += SessionTicketSize
	}

	copy(buf[off:], EncodeRefreshTicket(b.RefreshTicket))
	return buf
}

// DeserializeKeyBundle deserializes a KeyBundle.
func DeserializeKeyBundle(data []byte) (*KeyBundle, error) {
	if len(data) < 26 {
		return nil, errors.New("key bundle data too short")
	}

	b := &KeyBundle{}
	off := 0

	b.BundleGen = binary.BigEndian.Uint64(data[off:]); off += 8
	b.IssuedAtMs = binary.BigEndian.Uint64(data[off:]); off += 8
	b.ExpireAtMs = binary.BigEndian.Uint64(data[off:]); off += 8
	b.WorkerKID = binary.BigEndian.Uint16(data[off:]); off += 2

	// Policy
	b.Policy.TicketsPerBundle = data[off]; off++
	off++ // padding
	b.Policy.QueriesPerTicket = binary.BigEndian.Uint16(data[off:]); off += 2
	b.Policy.QueriesPerBundle = binary.BigEndian.Uint16(data[off:]); off += 2
	b.Policy.MaxClockSkewMs = binary.BigEndian.Uint32(data[off:]); off += 4
	b.Policy.AntiReplayWindow = binary.BigEndian.Uint16(data[off:]); off += 2
	b.Policy.TicketLifetimeMs = binary.BigEndian.Uint64(data[off:]); off += 8

	ticketCount := int(b.Policy.TicketsPerBundle)
	b.SessionTickets = make([]*SessionTicket, 0, ticketCount)
	for i := 0; i < ticketCount; i++ {
		if off+SessionTicketSize > len(data) {
			return nil, fmt.Errorf("not enough data for session ticket %d", i)
		}
		t, err := DecodeSessionTicket(data[off : off+SessionTicketSize])
		if err != nil {
			return nil, err
		}
		b.SessionTickets = append(b.SessionTickets, t)
		off += SessionTicketSize
	}

	if off+RefreshTicketSize > len(data) {
		return nil, errors.New("not enough data for refresh ticket")
	}
	rt, err := DecodeRefreshTicket(data[off : off+RefreshTicketSize])
	if err != nil {
		return nil, err
	}
	b.RefreshTicket = rt

	return b, nil
}
